circuitstats_test.go

```diff
@@ -3,41 +3,40 @@ package gnark
 import (
 	"encoding/gob"
 	"os"
+	"sync"
 	"testing"
 
 	"github.com/consensys/gnark-crypto/ecc"
 	"github.com/consensys/gnark/backend"
 	"github.com/consensys/gnark/frontend"
 	"github.com/consensys/gnark/internal/backend/circuits"
-	"github.com/stretchr/testify/require"
+	"github.com/consensys/gnark/test"
 )
 
 const (
 	fileStats        = "init.stats"
 	generateNewStats = false
 )
 
-func TestCircuitStatistics(t *testing.T) {
-
-	assert := require.New(t)
-
-	curves := ecc.Implemented()
-	for name, tData := range circuits.Circuits {
+var statsM sync.Mutex
 
-		for _, curve := range curves {
-			check := func(backendID backend.ID) {
-				t.Log(name, curve.String(), backendID.String())
-
-				ccs, err := frontend.Compile(curve, backendID, tData.Circuit)
-				assert.NoError(err)
-
-				// ensure we didn't introduce regressions that make circuits less efficient
-				nbConstraints := ccs.GetNbConstraints()
-				internal, secret, public := ccs.GetNbVariables()
-				checkStats(t, name, nbConstraints, internal, secret, public, curve, backendID)
+func TestCircuitStatistics(t *testing.T) {
+	assert := test.NewAssert(t)
+	for name := range circuits.Circuits {
+		for _, curve := range ecc.Implemented() {
+			for _, b := range backend.Implemented() {
+				// copy the circuit now in case assert calls t.Parallel()
+				tData := circuits.Circuits[name]
```

Comment:
--> Circuits store `frontend.Circuit` which is an interface, this is not copying the actual underlying value. 

Here I would probably just go for a lock in the internal/backend/TestCircuit data structure.

Reply:
ha, I see. What I saw on some user defined circuits is developers re-using circuits variable along the way, or doing some slice allocations in the Define. 
For example:
```
type Circuit struct {
A frontend.Variable
B []frontend.Variable
}

Define(...) {
circuit.A = api.Mul(circuit.A, circuit.A)
circuit.B = make([]frontend.Variable, ....)
}

```
which is definitely not thread safe --> 2 concurrent call to compile will not work.

Reply:
Do you think if trying to do it [here](https://github.com/ConsenSys/gnark/blob/master/test/assert.go#L371) would work? I.e. deep copy the circuit before first compilation and compare the circuit to the copied circuit after compilation?

Reply:
I found deep-copying circuits from their interface (frontend.Circuit) to be more complex than expected :) In internal/utils I believe, there is a ShallowCopy method.

Reply:
I see. I created a new issue for it https://github.com/ConsenSys/gnark/issues/195.